### ISSUE TITLE
Bug: no-unmodified-loop-condition fails to detect mutations inside arrow functions

### ISSUE THREAD

**Issue by synthex-byte** [eslint/eslint#21085]
### Environment

ESLint Playground

### What parser are you using?

Default (Espree)

### Playground Code
```js
let flag = false;

const cb = () => {
    flag = true;
};

while (flag === false) {
    cb();
}
```

### ESLint Configuration 
<details> 
 <summary> Tap to see full config </summary> 

```js
import { defineConfig } from "eslint/config";

export default defineConfig([
    {
        "rules": {
            "no-unmodified-loop-condition": [
                "error"
            ]
        }
    }
]);
```
 </details>

### ESLint Output
```
7:8 'flag' is not modified in this loop. (no-unmodified-loop-condition)
```

### What did you expect to happen?

No errors. flag is modified inside `cb`, and `cb()` is called in the loop body. The rule already handles this exact pattern when the callback is a function declaration:
```js
let flag = false;

function cb() {
    flag = true;
}

while (flag === false) {  // no error, as expected
    cb();
}
```
### What actually happened?

7:8 'flag' is not modified in this loop. (no-unmodified-loop-condition)

### Link to Minimal Reproducible Example

https://eslint.org/play/#eyJ0ZXh0IjoibGV0IGZsYWcgPSBmYWxzZTtcblxuY29uc3QgY2IgPSAoKSA9PiB7XG4gICAgZmxhZyA9IHRydWU7XG59O1xuXG53aGlsZSAoZmxhZyA9PT0gZmFsc2UpIHtcbiAgICBjYigpO1xufSIsIm9wdGlvbnMiOnsicnVsZXMiOnsibm8tdW5tb2RpZmllZC1sb29wLWNvbmRpdGlvbiI6WyJlcnJvciJdfSwibGFuZ3VhZ2VPcHRpb25zIjp7InBhcnNlck9wdGlvbnMiOnsiZWNtYUZlYXR1cmVzIjp7fX19fX0=

### Participation

- [x] I am willing to submit a pull request for this issue.

### AI acknowledgment

- [x] I did not use AI to generate this issue report.
- [ ] (If the above is not checked) I have reviewed the AI-generated content before submitting.

### Additional comments

This was actually raised back in 2016 in #7192, where @mysticatea confirmed the inconsistency and said "I think we can expand the support to function expressions and arrows". It was marked as an enhancement, but was closed after 21 days without reaching consensus. I'm happy to pick it up.

**Comment by kavixo** [comment]
This account was created just a few days ago, has no contribution history, and immediately began opening issues against ESLint. This has been happening a lot recently, and I'm not sure why. It might be worth it for the maintainers to look into GitHub's interaction limits or any settings that can restrict activity from newly created accounts, if something like that is available.

**Comment by nzakas** [comment]
This is an edge case we can't handle. It would involve tracking down every function call to see if it changed some variable and that would slow everything down. 

Our rules are meant to catch the 90% most common cases, and this falls outside of that range.
